### PR TITLE
change: use nospecialize for some calls for compilation and make sure…

### PR DESCRIPTION
**Pull request by kshyatt-aws**
… basis rotation instructions are counted

*Issue #, if available:* N/A

*Description of changes:* use `@nospecialize` for some methods to avoid unnecessary compilations. Also, make sure the qubits in `basis_rotation_instructions` are counted for `Program`s in case we have non-contiguous measurements.

*Testing done:* Unit tests passed locally.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/awslabs/braket-jl/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/awslabs/braket-jl/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/awslabs/braket-jl/blob/main/README.md) and [API docs](https://github.com/awslabs/braket-jl/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have installed and run [`git secrets`](https://github.com/awslabs/git-secrets) to make sure I did not commit any sensitive information (passwords or credentials)
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/circuit.jl [hunk 1]
```diff
@@ -271,6 +271,7 @@ QubitSet with 2 elements:
 qubits(c::Circuit) = (qs = union!(copy(c.moments._qubits), c.qubit_observable_set); QubitSet(qs))
 function qubits(p::Program)
     inst_qubits = mapreduce(ix->ix.target, union, p.instructions, init=Set{Int}())
+    bri_qubits  = mapreduce(ix->ix.target, union, p.basis_rotation_instructions, init=Set{Int}())
     res_qubits  = mapreduce(ix->(hasproperty(ix, :targets) && !isnothing(ix.targets)) ? reduce(vcat, ix.targets) : Set{Int}(), union, p.results, init=Set{Int}())
     return union(inst_qubits, res_qubits)
```
Comment (rmshaffer) [inline src/circuit.jl:276]: do you need to also return `bri_qubits`?

File: PyBraket/src/local_simulator.jl [hunk 2]
```diff
@@ -15,7 +15,7 @@ end
 Braket.simulate(d::PyLocalSimulator, task_spec::Braket.AnalogHamiltonianSimulation; shots::Int=0, kwargs...) = simulate(d, ir(task_spec); shots=shots, kwargs...)
 
 function Braket._run_internal(simulator::PyLocalSimulator, task_spec::AnalogHamiltonianSimulation, args...; kwargs...)
-    raw_py_result = simulator._run_internal(Py(ir(task_spec)), args...; kwargs...)
+    raw_py_result = simulator.run(Py(ir(task_spec)), args...; kwargs...).result()
```
Comment (rmshaffer) [inline PyBraket/src/local_simulator.jl:18]: 🥳

File: src/circuit.jl (hunk 1 above)
Comment (kshyatt-aws) [reply src/circuit.jl:276]: Good catch, yep

**Comment by kshyatt-aws** [comment]
The examples will be unhappy until we tag a new version

File: src/circuit.jl (hunk 1 above)
Comment (rmshaffer) [reply src/circuit.jl:276]: ok, I'll approve this, up to you whether you want to fix this in this PR or a follow-up

**Pull request merged**